File: python/03_largest_prime_factor.py
from math import sqrt

def largest_prime_factor(n: int) -> int:
    factor: int = 0
    root = int(sqrt(n))
    
    sieved = [False for _ in range(0, root + 1)]
    primes = []
    
    for i in range (2, len(sieved)):
        if not sieved[i]:
            for j in range (2*i, len(sieved), i):
                sieved[j] = True
            primes.append(i)
    
    for i in range(0, len(primes)):
        while n % primes[i] == 0:
            n //= primes[i]
            factor = primes[i]
    
    if n > 1:
        return n
    return factor
    
if __name__ == "__main__":
    print("Largest prime factor of 600851475143:", largest_prime_factor(600_851_475_143))

File: cpp/03_largest_prime_factor.cpp
#include <iostream>
#include <cmath>

int largest_prime_factor(unsigned long long n){
    int factor = 0;
    int len = std::sqrt(n) + 1;
    bool sieved[len] = {};
    for (int i = 2; i <= len; i++){
        if (!sieved[i]){
            for (int j = 2 * i; j <= len; j += i){
                sieved[j] = true;
            }
            if (n % i == 0){
                factor = i;
            }
        }
    }
    return factor;
}

int main() {
    std::cout << largest_prime_factor(600'851'475'143) << "\n";
}
